cmd/backup-manager/app/util/util.go

```diff
@@ -395,6 +400,313 @@ func GetBRMetaData(ctx context.Context, provider v1alpha1.StorageProvider) (*kvb
+	}
+
+	backupSize, err := calcBackupVolSnapshotSize(volSnapshots, snapshots)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return int64(backupSize), nil
+}
+
+// getSnapshotsFromBackupmeta read all snapshots from backupmeta
+// return volume - snapshot map
+func getSnapshotsFromBackupmeta(ctx context.Context, provider v1alpha1.StorageProvider) (map[string]string, error) {
+	newVolumeIDMap := make(map[string]string)
+
+	// read backup meta
+	s, err := util.NewStorageBackend(provider, &util.StorageCredential{})
+	if err != nil {
+		return newVolumeIDMap, err
+	}
+	defer s.Close()
+
+	var contents []byte
+	// use exponential backoff, every retry duration is duration * factor ^ (used_step - 1)
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Steps:    6,
+		Factor:   2.0,
+		Cap:      time.Minute,
+	}
+	readBackupMeta := func() error {
+		exist, err := s.Exists(ctx, constants.MetaFile)
+		if err != nil {
+			return err
+		}
+		if !exist {
+			return fmt.Errorf("%s not exist", constants.MetaFile)
+		}
+		contents, err = s.ReadAll(ctx, constants.MetaFile)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	isRetry := func(err error) bool {
+		return !strings.Contains(err.Error(), "not exist")
+	}
+	err = retry.OnError(backoff, isRetry, readBackupMeta)
+	if err != nil {
+		return nil, errors.Annotatef(err, "read backup meta from bucket %s and prefix %s", s.GetBucket(), s.GetPrefix())
+	}
+
+	metaInfo := &EBSBasedBRMeta{}
+	if err = json.Unmarshal(contents, metaInfo); err != nil {
+		return newVolumeIDMap, errors.Annotatef(err, "read backup meta from bucket %s and prefix %s", s.GetBucket(), s.GetPrefix())
+	}
+
+	// get volume-snapshot map
+	for i := range metaInfo.TiKVComponent.Stores {
+		store := metaInfo.TiKVComponent.Stores[i]
+		for j := range store.Volumes {
+			vol := store.Volumes[j]
+			newVolumeIDMap[vol.ID] = vol.SnapshotID
+		}
+	}
+
+	return newVolumeIDMap, nil
+}
+
+// getBackupVolSnapshots get a volue-snapshots map contains map[volumeId]{snapshot1, snapshot2, snapshot3}
+func getBackupVolSnapshots(volumes map[string]string) (map[string][]*ec2.Snapshot, error) {
+	volWithTheirSnapshots := make(map[string][]*ec2.Snapshot)
+
+	// read all snapshots from aws
+	ec2Session, err := NewEC2Session(CloudAPIConcurrency)
+	if err != nil {
+		klog.Errorf("new a ec2 session failure.")
+		return nil, err
+	}
+
+	// init search filter.Values
+	// init volWithTheirSnapshots
+	volValues := make([]*string, 0)
+	for volumeId := range volumes {
+		volValues = append(volValues, aws.String(volumeId))
+		if volWithTheirSnapshots[volumeId] == nil {
+			volWithTheirSnapshots[volumeId] = make([]*ec2.Snapshot, 0)
+		}
+	}
+
+	filters := []*ec2.Filter{{Name: aws.String("volume-id"), Values: volValues}}
+	// describe snapshot is heavy operator, try to call only once
+	// api has limit with max 1000 snapshots
+	// search with filter volume id the backupmeta contains
+	resp, err := ec2Session.ec2.DescribeSnapshots(&ec2.DescribeSnapshotsInput{
```

Comment:
I think we can put this request to the top in the `for` loop, so we can remove a duplicate request.

Reply:
From aws doc https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeSnapshots.html,  for the first call, we do not know what the `NextToken` is, and doc does not say can pass nil to it for the first call. so that keep it as a separate call (the first and later for loop).

Reply:
I see type of `NextToken` is string pointer, so its default value in struct is `nil`. In the codes here, you don't assign any value to `NextToken`, so the value is `nil`. You can just create a variable `var nextToken *string`, and assign it to `NextToken`. After the API call, you assign `resp.NextToken` to `nextToken`. Then all the API requests have right token.